### DISCUSSION TITLE
Function is not being called in switch clause

### DISCUSSION THREAD

**Discussion by ogomez92** [savonet/liquidsoap#2972] [User Support]
**Describe the bug**
A clear and concise description of what the bug is. Add some script and log extract so we can see what is happening. Don't forget to add some context although not necessarily your whole script..
First of all I'm not sure this is a bug and there might be something I'm missing but I don't think it's the case because I'm passing a function call in the switch clause as you will see in the code.

I am trying to use Liquidsoap and adding time announcements to my station.
Every time this code runs, it correctly adds an announcement every hour, but it adds the same file every time e.g. If I run the script at 19:50, it will play 08.mp3 for 20:00 at the correct time, but then at 21:00, instead of playing 09.mp3 it will play 08.mp3 again.

Is there something I am missing in this script?
I tried to find the solution in different sources but the liquidsoap APIdocs don't give many examples that I've seen and there aren't many instructions out there.

Any help is appreciated.

```Liquidsoap
def play_announcement()
current_hour = string.trim(system("date +'%I'").stdout)

  announcement_file = "/home/radio/horas/"^ current_hour ^ ".mp3"
  announcement = single(announcement_file)

  announcement
end

radio=mksafe(eat_blank(playlist("/home/radio/songs")))
radio_with_announcement = add(weights=[0.25, 0.75], [radio, switch([({0m0s},play_announcement())])])

full =
  fallback(track_sensitive=false,[
    input.http("http://localhost:8000/livesource.mp3"),
  radio_with_announcement
])

output.icecast(%mp3.cbr(bitrate=160),
id="ice",
host="127.0.0.1",port=8000,password="xxxxxx",
mount="stream.mp3",full)
```
**To Reproduce**
A minimal script to reproduce the issue is the most efficient way to have your bug addressed and fixed very quickly!
You can run the code above changing 0m0s for 0s in the switch and you can get the time announcements from
https://oriolgomez.com/files/horas.zip

**Expected behavior**
It should play the correct file at the correct time when a function reference is passed.
**Version details**
2.1.1
- OS: [e.g. Debian, OSX]
- Version [e.g. `1.3.4`]
Ubuntu 22
**Install method**
Did you install via `opam`? Using your distribution's packages? From source?
opam
**Common issues**
Can't decode or encode in some audio format? Check if you have [installed the correct dependencies](https://www.liquidsoap.info/#download)

**Comment by toots** [comment]
Hi!

Your issue is because the function `play_announcement` is executed once when the script starts and never again so the source is created with whatever hour there is and never updated.

Also your time predicate is pretty tight, one second after and it will fail!

The best I believe would be to push a request per hour and have it ready in advance. You can use `thread.when` for that. Let's try something. I'm writing this script without testing it but it should be close to what you need.

```liquidsoap
# Top of the hour queue
top_of_the_hour = request.queue(id="top_of_the_hour")

# When we're at minute `30`, queue a new song to make sure it is fully prepared at the top the next hour
def queue_announcement()
  next_hour = time.local().hour+1

  announcement_file = "/home/radio/horas/#{next_hour}.mp3"

  request = request.create(announcement_file)

  top_of_the_hour.push(request)
end

thread.when({30m}, queue_announcement)

# Now create a source that will be ready and play a single request at the top of each hour:
top_of_the_hour = switch([
  ({0m}, top_of_the_hour)
])
```
